[WF] Support skipping merge with base if PR is relatively recent (#12426)
buildbuddy-io/buildbuddy: docs/workflows-config.md

    # ...
```

> ## Merge with base

Reviewer: Somewhere in this section, should we mention how this relates to merge queues? I guess something like: even if you are using merge queues, merge_with_base can help catch issues with a PR even sooner, before it reaches the merge queue (?)

Author: Added a section

By default, when workflows are triggered by `pull_request` events, the CI runner will merge
the PR branch with the PR's base branch (i.e. main). This is controlled by
`merge_with_base`, which defaults to `true`.

This may help catch integration problems before merge,
such as merge conflicts or tests that only fail when the PR is combined
with recent changes on the base branch.

```yaml title="buildbuddy.yaml"
actions:
  - name: "Test"
    triggers:
      pull_request:
        branches: ["main"]
        merge_with_base: true # default
    steps:
      - run: "bazel test //..."
```

This behavior may hurt CI performance, because the base branch must be
fetched on every run, and unrelated base branch changes that were merged
into the git workspace may invalidate the Bazel cache and require rebuilds.
This behavior can be disabled by setting `merge_with_base: false` or by setting
`max_base_staleness_before_merge`.

When `max_base_staleness_before_merge` is set, the CI runner will only merge the base branch into the PR branch when the PR's merge base is at least this stale relative to the base branch tip.

> For example, if `max_base_staleness_before_merge: "24h"`, the main branch tip is 1h old, and the merge base between

Reviewer: I think people probably see this 24h example and assume it's our recommended default.

Should we maybe start with something shorter like 3h or 6h maybe, since the higher this value, the more cache eviction churn we will expect to see (iiuc?) since (e.g.) if somebody upgrades their Go toolchain or something like that, there will be a 24h window where we keep two copies of the toolchain and all of its transitive deps fresh in cache, vs just the latest one always.

Author: Removed reference to 24h. Added a 3h cap to the interval, and used that value in the docs

the PR and main branch is 2h old, the PR is not considered stale and should not be merged with main.
If the merge base was 26h old and 25h behind the main branch tip, the PR

> would be considered stale and should be merged with main.

Reviewer: Once the 24h elapses, it seems like we transition to a state where we always merge with the latest commits on master, until the user pushes a new commit to their PR branch. Is that right?

If so, would there be an edge case where if someone re-runs a workflow on an old PR branch, it won't be merged w/ master and potentially trigger a bunch of cache uploads for artifacts that have already expired / are no longer relevant on the main branch?

I wonder if the behavior should be something like: every time we merge with the base branch, add a local git ref like `buildbuddy/last-merged-base-commit/<base-branch-name>` or something like that, and then we always merge with that commit instead of the base branch tip commit, until that new commit becomes 24h old, and then we bump the last-merged-base-commit to `HEAD`.

Author: With this PR, once 24h elapses, we'd always merge with master until they rebase their PR branch. I agree it's not ideal how users might accidentally fall into this situation without realizing

For your suggestion, are you suggesting that for every PR, we merge with the current master tip the first time the PR workflow runs. For subsequent runs on the same branch, we only merge again if 24h hours have passed. A run on a different PR branch would pull in a different master tip

Or are you suggesting that we share the same `buildbuddy/last-merged-base-commit/<base-branch-name>` across all PRs?

Author: I think Option A is a good idea. Option B would be hard with our snapshot sharing strategy, because workflows on PR branches don't save master snapshots

Option A might seem a little flaky/unpredictable though, because most PR workflows only write local snapshots. If the workflow is assigned to a different executor, it will merge with base anew

Author: I have a new idea that is both more deterministic and more stable.

We merge with base with master from midnight of the current day. That way we don't have to store the merge base in a local git ref that we can't always guarantee exists

Ex.
Workflow runs at Tuesday at 9AM. We merge with master @ Monday midnight.
Workflow runs at Tuesday 3PM. We merge with master @ Monday midnight. The merge base is the same, so there are lots of cache hits
Workflow runs at Wed. 9AM. We merge with master @ Tuesday midnight. There will be a different merge base this time, but we're keeping the workflow relatively up-to-date with master

Customers could customize the timeframe, and we always take the floor of that duration. If they set 24h, the window starts at midnight. If they set 6h, the windows start at [midnight, 6AM, noon, 6PM]. etc. 

If the PR's merge base is newer than the calculated default merge base, we always skip the merge with base

@bduffany WDYT?

Author: Updated to use this ^^

### Merge with base interval

When `merge_with_base_interval` is set, the CI runner merges with the oldest base branch commit within the interval,
rather than the base branch tip.

The intention is that multiple CI runs within an interval will all merge with the same commit, even
if the base branch tip has advanced. Repeated runs of an unchanged PR will produce the same merged result and hit a warm Bazel cache.

The runner will still merge with the base branch tip once per interval, keeping the PR reasonably up to date with the base branch to catch integration issues.

If the PR's merge base is already newer than the compute commit, the merge is skipped.
And if the interval contains no base branch commits, it falls back to the tip.

For example, with `merge_with_base_interval: "3h"`, the base advances at 00:00, 03:00,
06:00, ... UTC, and all runs within the same 3h window merge with the first base
branch commit after the window boundary:

- The interval boundary is [09:00 UTC, 12:00 UTC].
- Main is pushed at 08:00 UTC (commit ABC). This commit is outside the current interval.
- Main is pushed at 09:05 UTC (commit DEF). This commit is the oldest commit in the current interval.

> - Main is pushed at 10:00 UTC (commit GHI). This commit is in the current interval, but is not the oldest commit in the interval..

Reviewer: ```suggestion
- Main is pushed at 10:00 UTC (commit GHI). This commit is in the current interval, but is not the oldest commit in the interval.
```

- Main is pushed at 12:05 UTC (commit JKL). This commit is in the next interval.
- A PR Workflow runs at 09:30 UTC. It merges with the oldest base branch commit in the current interval (commit DEF).
- A PR Workflow runs at 10:45 UTC. Even though Main has advanced to commit GHI, it merges with the oldest base branch commit in the current interval (commit GHI).
  This lets it reuse the cache from the earlier PR run.
- A run at 12:15 UTC is in the next interval, so it merges with commit JKL.

>   This helps keep the PR reasonably up to date with the base branch.

Reviewer: ultra nit: `jkl` / `ghi` aren't valid commit SHA characters (0-9 / a-f) 🤡  

```suggestion
- Main is pushed at 08:00 UTC (commit `a`). This commit is outside the current interval.
- Main is pushed at 09:05 UTC (commit `b`). This commit is the oldest commit in the current interval.
- Main is pushed at 10:00 UTC (commit `c`). This commit is in the current interval, but is not the oldest commit in the interval..
- Main is pushed at 12:05 UTC (commit `d`). This commit is in the next interval.
- A PR Workflow runs at 09:30 UTC. It merges with the oldest base branch commit in the current interval (commit `b`).
- A PR Workflow runs at 10:45 UTC. Even though Main has advanced to commit `c`, it merges with the oldest base branch commit in the current interval (commit `c`).
  This lets it reuse the cache from the earlier PR run.
- A run at 12:15 UTC is in the next interval, so it merges with commit `d`.
  This helps keep the PR reasonably up to date with the base branch.
```

Author: Applied

> With a shorter interval like `1h`, the base advances at the top of each hour.

Reviewer: ```suggestion
With a shorter interval like `1h`, the base advances at the start of each hour.
```

`merge_with_base_interval` is capped at `3h` so that PRs are not merged with an overly
stale base, which would also have the negative side effect of keeping stale
artifacts in the cache longer. Use a shorter interval when merge confidence is
more important, and a longer interval when cache stability and reduced CI churn
are more important.

### Interaction with merge queues

Merge with base is still useful when your repo uses merge queues. Merging with base on
`pull_request` workflows can catch PR-specific issues earlier, before the PR
reaches the merge queue.

> Workflows that run on merge queue branches themselves
> do not need `merge_with_base`, since the merge queue commit should already be
> merged with the base branch.

Reviewer: > Workflows that run on merge queue branches themselves do not need `merge_with_base`

idk if this would be obvious to readers or not, but in addition to being unnecessary, it's also unsupported. Here is some alternate phrasing, but not sure if it's any better

```suggestion
Note that merge queue workflows can only be triggered on `push` events, where `merge_with_base` is not supported. The merge queue commit already contains the merged base branch and the changes in the merge queue.
```

Author: I decided to keep both sentences to include the most info

    # ...
```

## Merge with base

By default, when workflows are triggered by `pull_request` events, the CI runner will merge

> the PR branch with the PR's base branch (i.e. main). This is controlled by

Reviewer: ```suggestion
the PR branch with the PR's base branch (e.g. main). This is controlled by
```

`merge_with_base`, which defaults to `true`.

This may help catch integration problems before merge,
such as merge conflicts or tests that only fail when the PR is combined
with recent changes on the base branch.

```yaml title="buildbuddy.yaml"
actions:
  - name: "Test"
    triggers:
      pull_request:
        branches: ["main"]
        merge_with_base: true # default
    steps:
      - run: "bazel test //..."
```

By default, the PR is merged with the current base branch tip. This may hurt CI
performance if the base branch tip advances frequently. Merging in
unrelated base branch changes may invalidate the Bazel cache and require
rebuilds, even when the PR branch itself is unchanged. This behavior can be
disabled by setting `merge_with_base: false`, or made less frequent by setting
`merge_base_interval`.

### Merge base interval

> When `merge_base_interval` is set, instead of merging with the base branch tip,

Reviewer: A "merge base" is a specific concept in git so I think `merge_base_interval` might be unclear on its own. Maybe fully write out `merge_with_base_interval` which also feels a bit more consistent with the feature that it's controlling (`merge_with_base`) ?

> the CI runner merges with the oldest base branch commit in the

Reviewer: Took me a while to fully understand this sentence:
- "Oldest" could refer to a commit that is even older than the common `merge-base` commit between the PR branch and its base branch
- On first thought, it seems surprising that we would be going "backwards" and merging with an even older commit than the `merge-base`
- On second thought, would we still pick up all commits up to and including the `merge-base`? (is that how it works?)

Author: I tried to clarify the docs. PTAL

the Mac executors in the `workflows` pool should execute the
workflow, and BuildBuddy will publish the results to your branch.

> ## Debugging

Reviewer: ```suggestion
## Troubleshooting
```
